### PR TITLE
Add throughput test

### PR DESCRIPTION
**Pull request by ppalucha**
New throughput-test: true property allows to run a test that uses
multiple threads to run queries in a round-robin fashion, without
any waits between queries.

Each thread runs a different permutation of queries, as described
in appendix D of TPC-DS specification at
http://www.tpc.org/tpc_documents_current_versions/pdf/tpc-ds_v3.2.0.pdf.

Pre-warm runs are run in parallel but runs are spreaded between all
threads.

The intended usage is to:
- specify all tpc-ds queries in query-names property
- use small schema (e.g. sf10)
- set concurrency to expected number of threads
- use high number of runs
- specify time limit for test, so benchmark ends after limit is
reached.

We report only queries that finish before time limit is reached.

To maximise throughput, query metrics are not fetched from Trino. Only
number of successful and failed queries is reported for each thread.

### CODE REVIEW

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 1]
```diff
@@ -150,52 +158,145 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
...
+    private List<QueryExecutionResult> executeConcurrentQueries(Benchmark benchmark, int runs, boolean reportStatus, Optional<ZonedDateTime> lastTime, int threadNumber, Integer[] queryOrder)
+            throws SQLException
+    {
+        boolean firstQuery = true;
+        List<QueryExecutionResult> queryExecutionResults = newArrayList();
+        try (Connection connection = getConnectionFor(new QueryExecution(benchmark, benchmark.getQueries().get(0), 0))) {
+            for (int run = 1; run <= runs; run++) {
+                for (int queryIndex = 0; queryIndex < benchmark.getQueries().size(); queryIndex++) {
+                    int permutedQueryIndex = queryIndex;
+                    if (!reportStatus) {
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:217]: don't we want to report status always?

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 2]
```diff
@@ -150,52 +158,145 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
...
+                        LOG.info("Comparing {} vs {}, {}, {}", queryIndex % benchmark.getConcurrency(), threadNumber, queryIndex, benchmark.getConcurrency());
+                        if (queryIndex % benchmark.getConcurrency() != threadNumber) {
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:219]: does pre-warm account for total query time? Is pre-warm a separate stage before actual run?

File: benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class PermutationUtils
+{
+    public static Integer[] preparePermutation(int size, int permutationId)
+    {
+        List<Integer> intList = IntStream.range(0, size).boxed().collect(Collectors.toList());
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java:26]: static import `toList` here (simialr in other places that use `toList` or `toImmutableList`)

File: benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        Collections.shuffle(intList, new Random(permutationId * 7823L));
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java:27]: can you just put `private static final RANDOM = new Random()`

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java [hunk 5]
```diff
@@ -167,6 +171,38 @@ private void checkClocksSync()
                 });
     }
 
+    @Override
+    public Future<?> concurrencyTestExecutionFinished(List<QueryExecutionResult> executions)
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:175]: do you know, if benchto-server UI still works after these changes and reported benchmarks?

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 6]
```diff
@@ -64,15 +72,15 @@
     @Autowired
     private ApplicationContext applicationContext;
 
-    public BenchmarkExecutionResult execute(Benchmark benchmark, int benchmarkOrdinalNumber, int benchmarkTotalCount)
+    public BenchmarkExecutionResult execute(Benchmark benchmark, int benchmarkOrdinalNumber, int benchmarkTotalCount, Optional<ZonedDateTime> lastTime)
```
[the remark this answers is not in the document]
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java]: Renamed to executionTimeLimit

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 7]
```diff
@@ -150,52 +158,145 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
...
                 QueryExecution queryExecution = new QueryExecution(benchmark, query, run);
-
                 executionCallables.add(() -> {
-                    QueryExecutionResult result;
-
+                    List<QueryExecutionResult> queryExecutionResults = new ArrayList<>();
                     try (Connection connection = getConnectionFor(queryExecution)) {
-                        macroService.runBenchmarkMacros(benchmark.getBeforeExecutionMacros(), benchmark, connection);
+                        Optional<QueryExecutionResult> result = executeSingleQuery(queryExecution, benchmark, connection, reportStatus, lastTime);
+                        result.ifPresent(queryExecutionResults::add);
```
[the remark this answers is not in the document]
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:183]: Empty result is returned if query exceeds executionTimeLimit. This may happen also for standard query.

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 8]
```diff
@@ -150,52 +158,145 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
...
+    private List<Callable<List<QueryExecutionResult>>> buildConcurrencyQueryExecutionCallables(Benchmark benchmark, int runs, boolean reportStatus, Optional<ZonedDateTime> lastTime)
+    {
+        List<Callable<List<QueryExecutionResult>>> executionCallables = newArrayList();
+        for (int thread = 0; thread < benchmark.getConcurrency(); thread++) {
+            int finalThread = thread;
+            executionCallables.add(() -> {
+                LOG.info("Running throughput test: {} queries, {} runs", benchmark.getQueries().size(), runs);
+                Integer[] queryOrder = PermutationUtils.preparePermutation(benchmark.getQueries().size(), finalThread);
+                List<QueryExecutionResult> queryExecutionResults = executeConcurrentQueries(benchmark, runs, reportStatus, lastTime, finalThread, queryOrder);
+                if (reportStatus && !queryExecutionResults.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:201]: None of the measured queries finished before time limit.

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java (hunk 1 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:219]: False reportStatus basically means we are doing pre-warming runs.

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java (hunk 2 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:220]: Pre-warm is a separate stage, results of pre-warm runs are not reported - see https://github.com/ppalucha/benchto/blob/pp/concurrency-round-robin-test/benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java#L107

For "standard" benchmark usually each benchmark contains only single query, therefore pre-warming runs for specific query are run just before this query. For concurrency tests, benchmark contains multiple queries. Pre-warming runs for all queries are executed before measured executions.

File: benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class PermutationUtils
+{
+    public static Integer[] preparePermutation(int size, int permutationId)
```
[the remark this answers is not in the document]
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java:24]: Changed to int[].

File: benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java (hunk 4 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java:27]: I'm not sure what you mean -  I need permutations to be constant for given thread number to assure tests are repeatable.

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java (hunk 5 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:175]: It displays an execution, but doesn't show any details / metrics. I can check what is missing and if it's easy to get some proper output displayed in UI.

**Review by przemekak — APPROVED** [review]

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 10]
```diff
@@ -134,13 +142,15 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
...
         ListeningExecutorService executorService = executorServiceFactory.create(benchmark.getConcurrency());
         try {
-            List<Callable<QueryExecutionResult>> queryExecutionCallables = buildQueryExecutionCallables(benchmark, runs, reportStatus);
-            List<ListenableFuture<QueryExecutionResult>> executionFutures = (List) executorService.invokeAll(queryExecutionCallables);
-            return Futures.allAsList(executionFutures).get();
+            List<Callable<List<QueryExecutionResult>>> queryExecutionCallables = buildQueryExecutionCallables(benchmark, runs, reportStatus, executionTimeLimit);
+            List<ListenableFuture<List<QueryExecutionResult>>> executionFutures = (List) executorService.invokeAll(queryExecutionCallables);
+            return Futures.allAsList(executionFutures).get().stream()
+                    .flatMap(List::stream)
+                    .collect(ImmutableList.toImmutableList());
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:153]: static import `toImmutableList`

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sopel39) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:183]: > Empty result is returned if query exceeds executionTimeLimit. This may happen also for standard query.

Exception isn't throw here?

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java (hunk 2 above)
Comment (sopel39) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:220]: OK, but are all threads waiting for pre-warm to complete before running actual queries?

**Review by sopel39 — COMMENTED** [review]
Would it be possible to split this into two commits:
1. refactor `BenchmarkExecutionDriver` (extract methods)
2. add concurrent testing?

**Review by sopel39 — COMMENTED** [review]
is it possible to add tests to `BenchmarkExecutionDriverTest` or `ExecutionDriverTest`?

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:183]: At lowest level I throw timeout exception, but then I handle it by returning empty query result.

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java (hunk 2 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:220]: Yes, they wait.

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 11]
```diff
@@ -150,52 +160,144 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
...
+
+        if (reportStatus) {
+            statusReporter.reportExecutionStarted(queryExecution);
+        }
+        QueryExecutionResultBuilder failureResult = new QueryExecutionResultBuilder(queryExecution)
+                .startTimer();
+        try {
+            result = queryExecutionDriver.execute(queryExecution, connection);
+            if (isTimeLimitExceeded(executionTimeLimit)) {
+                throw new TimeoutException(format("Execution of query %s for benchmark %s exceeded time limit",
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:269]: why not return `Optional.empty()` here directly?

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/ExecutionDriver.java [hunk 12]
```diff
@@ -145,6 +145,12 @@ private boolean isTimeLimitEnded()
         return timeLimit.isPresent() && timeLimit.get().compareTo(Duration.between(startTime, nowUtc())) < 0;
     }
 
+    private Optional<ZonedDateTime> getLastTime()
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/ExecutionDriver.java:148]: rename

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/ExecutionDriver.java [hunk 13]
```diff
@@ -124,15 +124,15 @@ private void executeBenchmarks(List<Benchmark> benchmarks)
...
+            benchmarkExecutionResults.add(benchmarkExecutionDriver.execute(benchmark, benchmarkOrdinalNumber++, benchmarks.size(), getLastTime()));
             benchmarkStatusReporter.processCompletedFutures();
         }
 
         List<BenchmarkExecutionResult> failedBenchmarkResults = benchmarkExecutionResults.stream()
                 .filter(benchmarkExecutionResult -> !benchmarkExecutionResult.isSuccessful())
                 .collect(toList());
 
-        benchmarkStatusReporter.awaitAllFutures(10, TimeUnit.MINUTES);
+        benchmarkStatusReporter.awaitAllFutures(20, TimeUnit.MINUTES);
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/ExecutionDriver.java:135]: why?

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java [hunk 14]
```diff
@@ -167,6 +171,38 @@ private void checkClocksSync()
...
+            FinishRequest finishRequest = new FinishRequestBuilder()
+                    .withStatus(ENDED)
+                    .withEndTime(
+                            executions.stream()
+                                    .map(e -> e.getUtcEnd().toInstant())
+                                    .max(Comparator.comparing(Instant::toEpochMilli))
+                                    .orElseThrow(NoSuchElementException::new))
+                    .addMeasurement(Measurement.measurement(
+                            "queries_successful",
+                            "NONE",
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:187]: `QUERY_COUNT` instead of `NONE`

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java [hunk 15]
```diff
@@ -167,6 +171,38 @@ private void checkClocksSync()
...
+                            executions.stream().filter(QueryExecutionResult::isSuccessful).count()))
+                    .addMeasurement(Measurement.measurement(
+                            "queries_failed",
+                            "NONE",
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:191]: `QUERY_COUNT` instead of `NONE`

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java [hunk 16]
```diff
@@ -167,6 +171,38 @@ private void checkClocksSync()
...
+                            executions.stream().filter(query -> !query.isSuccessful()).count()))
+                    .addAttribute(
+                            "queries_order",
+                            executions.stream().limit(99).map(QueryExecutionResult::getQueryName).collect(Collectors.joining(",")))
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:195]: put `.limit`, `.collect` and `.map` in separate lines

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java [hunk 17]
```diff
@@ -167,6 +171,38 @@ private void checkClocksSync()
...
+                                    .orElseThrow(NoSuchElementException::new))
+                    .addMeasurement(Measurement.measurement(
+                            "queries_successful",
+                            "NONE",
+                            executions.stream().filter(QueryExecutionResult::isSuccessful).count()))
+                    .addMeasurement(Measurement.measurement(
+                            "queries_failed",
+                            "NONE",
+                            executions.stream().filter(query -> !query.isSuccessful()).count()))
+                    .addAttribute(
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:193]: why limit it to 99?

**Comment by ppalucha** [comment]
> is it possible to add tests to `BenchmarkExecutionDriverTest` or `ExecutionDriverTest`?

Current tests are only mocks, they do not run executions, so there is no change to test for concurrency tests. 

> Would it be possible to split this into two commits:
> 
> 1. refactor `BenchmarkExecutionDriver` (extract methods)
> 2. add concurrent testing?

I split into 2 commits.

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/ExecutionDriver.java (hunk 13 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/ExecutionDriver.java:135]: right, not needed

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java (hunk 14 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:191]: There's no such unit as QUERY_COUNT (they are defined in benchto-service/src/main/java/io/trino/benchto/service/model/MeasurementUnit.java)

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java (hunk 15 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:195]: As above - no such unit.

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java (hunk 17 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:197]: removed

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java]: Removed exception - I just return empty result on timeout.

File: benchto-driver/src/test/java/io/trino/benchto/driver/utils/PermutationUtilsTest.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+ * limitations under the License.
+ */
+package io.trino.benchto.driver.utils;
+
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.stream.IntStream;
+
+import static org.junit.Assert.assertArrayEquals;
```
Comment (sopel39) [inline benchto-driver/src/test/java/io/trino/benchto/driver/utils/PermutationUtilsTest.java:21]: is there `assertArrayNotEquals`?

File: benchto-driver/src/test/java/io/trino/benchto/driver/utils/PermutationUtilsTest.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+public class PermutationUtilsTest
+{
+    @Test
+    public void preparePermutation()
+    {
+        int[] permutation1 = PermutationUtils.preparePermutation(10, 1);
+        int[] permutation2 = PermutationUtils.preparePermutation(10, 2);
+        int[] permutation3 = PermutationUtils.preparePermutation(10, 2);
+        int[] sorted = IntStream.range(0, 10).toArray();
+        assertFalse(Arrays.equals(permutation1, permutation2));
```
Comment (sopel39) [inline benchto-driver/src/test/java/io/trino/benchto/driver/utils/PermutationUtilsTest.java:33]: can you use `assertThat(permutation1).isNotEqualTo(permutation2)`?

File: benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.Random;
+import java.util.stream.IntStream;
+
+import static java.util.stream.Collectors.toList;
+
+public class PermutationUtils
+{
+    public static int[] preparePermutation(int size, int permutationId)
+    {
+        List<Integer> intList = IntStream.range(0, size).boxed().collect(toList());
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java:27]: move `.range` `.boxed` and `toList` to separate lines

File: benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+        Collections.shuffle(intList, new Random(permutationId * 7823L));
+        return intList.stream().mapToInt(i->i).toArray();
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/utils/PermutationUtils.java:29]: I think you need to reformat code (should be `i -> i`)

File: benchto-driver/src/main/java/io/trino/benchto/driver/loader/BenchmarkLoader.java [hunk 22]
```diff
@@ -413,6 +415,6 @@ private String createFormatString(Collection<Benchmark> benchmarks)
         int nameMaxLength = benchmarks.stream().mapToInt((benchmark) -> benchmark.getName().length()).max().orElseGet(() -> 10);
         int dataSourceMaxLength = benchmarks.stream().mapToInt((benchmark) -> benchmark.getDataSource().length()).max().orElseGet(() -> 10);
         int indent = 3;
-        return "\t| %-" + (nameMaxLength + indent) + "s | %-" + (dataSourceMaxLength + indent) + "s | %-4s | %-8s | %-11s |";
+        return "\t| %-" + (nameMaxLength + indent) + "s | %-" + Math.max(dataSourceMaxLength + indent, 11) + "s | %-4s | %-8s | %-11s | %-15s |";
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/loader/BenchmarkLoader.java:418]: nit: `Math.max(dataSourceMaxLength + indent, 11)` should be separate commit

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java [hunk 23]
```diff
@@ -179,6 +221,7 @@ private FinishRequest buildExecutionFinishedRequest(QueryExecutionResult executi
         if (executionResult.getPrestoQueryId().isPresent()) {
             requestBuilder.addAttribute("prestoQueryId", executionResult.getPrestoQueryId().get());
         }
+        requestBuilder.addAttribute("queryName", executionResult.getQueryExecution().getQueryName());
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:224]: Benchto should already have `queryName`, why we need this?

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 24]
```diff
@@ -196,12 +275,17 @@ private QueryExecutionResult executeSingleQuery(
...
 
     private Connection getConnectionFor(QueryExecution queryExecution)
             throws SQLException
     {
         return applicationContext.getBean(queryExecution.getBenchmark().getDataSource(), DataSource.class).getConnection();
     }
+
+    private boolean isTimeLimitExceeded(Optional<ZonedDateTime> executionTimeLimit)
+    {
+        return executionTimeLimit.isPresent() && executionTimeLimit.get().compareTo(nowUtc()) < 0;
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:289]: use:
```
executionTimeLimit.ifPresent(limit -> ...).orElse(false);
```

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 25]
```diff
@@ -196,12 +275,17 @@ private QueryExecutionResult executeSingleQuery(
...
+        return Optional.of(result);
     }
 
     private Connection getConnectionFor(QueryExecution queryExecution)
             throws SQLException
     {
         return applicationContext.getBean(queryExecution.getBenchmark().getDataSource(), DataSource.class).getConnection();
     }
+
+    private boolean isTimeLimitExceeded(Optional<ZonedDateTime> executionTimeLimit)
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:287]: nit: time linit change should be a separate commit

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 26]
```diff
@@ -158,19 +173,80 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
...
+                    }
+                    Optional<QueryExecutionResult> result = executeSingleQuery(queryExecution, benchmark, connection, false, executionTimeLimit);
+                    result.ifPresent(queryExecutionResults::add);
+                }
+            }
+        }
+        return queryExecutionResults;
+    }
+
+    private Optional<QueryExecutionResult> executeSingleQuery(
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:244]: Just throw some `TimeLimitException extends RuntimeException` instead of returning `Optional`.
`Optional` is kind of meaningless (was it OK or error?, what caused error?).

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 27]
```diff
@@ -158,19 +173,80 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
...
+                    }
+                    else {
+                        permutedQueryIndex = queryOrder[queryIndex];
+                    }
+                    Query query = benchmark.getQueries().get(permutedQueryIndex);
+                    if (isTimeLimitExceeded(executionTimeLimit)) {
+                        LOG.warn("Interrupting benchmark {} due to time limit exceeded", benchmark.getName());
+                        return queryExecutionResults;
+                    }
+                    QueryExecution queryExecution = new QueryExecution(benchmark, query,
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:228]: put each argument in single line

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java [hunk 28]
```diff
@@ -158,19 +173,80 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
...
+                            queryIndex
+                                    + threadNumber * benchmark.getQueries().size()
+                                    + (run - 1) * benchmark.getConcurrency() * benchmark.getQueries().size());
+                    if (firstQuery && reportStatus) {
+                        statusReporter.reportExecutionStarted(queryExecution);
```
Comment (sopel39) [inline benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:233]: why do we report first query?

**Review by sopel39 — APPROVED** [review]
lgtm % comments

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java (hunk 28 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:233]: We report only single execution per benchmark, so we report start of the first query and finish of all the queries. This is because:
- There's nothing really interesting to report for single queries.
- Benchto has troubles keeping up with rate of report requests, it caused growing backlog of futures to execute.

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java (hunk 25 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java:287]: Currently time limit is only used in concurrency benchmark (for standard we pass empty Optional as time limit). So this would need to be a strange no-op change to introduce it separately.

File: benchto-driver/src/test/java/io/trino/benchto/driver/utils/PermutationUtilsTest.java (hunk 18 above)
Comment (ppalucha) [reply benchto-driver/src/test/java/io/trino/benchto/driver/utils/PermutationUtilsTest.java:21]: nope...

File: benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java (hunk 23 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/listeners/BenchmarkServiceExecutionListener.java:224]: Current queryName is a variable on benchmark level - it would make more sense to have is on attribute level. But it's a leftover from earlier version of code when I reported all executions - I will remove it.

File: benchto-driver/src/test/java/io/trino/benchto/driver/utils/PermutationUtilsTest.java (hunk 18 above)
Comment (ppalucha) [reply benchto-driver/src/test/java/io/trino/benchto/driver/utils/PermutationUtilsTest.java:21]: never mind, will use assertThat() anyway

File: benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java (hunk 24 above)
Comment (ppalucha) [reply benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java]: Actually, had to use map() as ifPresent doesn't return value.
